### PR TITLE
[stable8.2] Catch exception when querying direct download link

### PR DESCRIPTION
**Pull request by PVince81**
Potential fix for the 8.2 part of https://github.com/owncloud/core/issues/21033

@LukasReschke

### CODE REVIEW

File: lib/private/connector/sabre/filesplugin.php [hunk 1]
```diff
@@ -188,9 +189,14 @@ public function handleGetProperties(PropFind $propFind, \Sabre\DAV\INode $node)
...
 				/** @var $node \OC\Connector\Sabre\File */
-				$directDownloadUrl = $node->getDirectDownload();
-				if (isset($directDownloadUrl['url'])) {
-					return $directDownloadUrl['url'];
+				try {
+					$directDownloadUrl = $node->getDirectDownload();
+					if (isset($directDownloadUrl['url'])) {
+						return $directDownloadUrl['url'];
+					}
+				} catch (StorageNotAvailableException $e) {
```
Comment (MorrisJobke) [inline lib/private/connector/sabre/filesplugin.php]: Fetching all the StorageNotAvailableExceptions without saying anything. Is this okay? Or should we log something?

Comment (PVince81) [inline lib/private/connector/sabre/filesplugin.php]: Not in this case. If the client does a PROPFIND directly on this file/folder, they will get a real 503 HTTP error.
The purpose of this one here is to make sure the file list can still be fetched by the client even if one entry is from an unavailable storage

File: lib/private/connector/sabre/filesplugin.php [hunk 2]
```diff
@@ -188,9 +189,14 @@ public function handleGetProperties(PropFind $propFind, \Sabre\DAV\INode $node)
...
+					// return empty download link when storage not available
+					return null;
```
Comment (PVince81) [inline lib/private/connector/sabre/filesplugin.php]: in the new Sabre I needed to return false here, so let me see if I can do the same to make this backport closer

**Comment by PVince81** [comment]
Ok, updated the PR to return false instead of null.

**Comment by PVince81** [comment]
Please review @MorrisJobke @icewind1991 @nickvergessen.

This is the backport of https://github.com/owncloud/core/pull/21087

- [ ] Approval pending https://github.com/owncloud/core/pull/21087#issuecomment-163692949

**Comment by MorrisJobke** [comment]
I can't reproduce the error:

* server 1 has stable8.2 
* server 2 has stable8.0
* server 2 shares a file to a user on server 1
* user on server 1 accepts the remote share
* user on server 2 deletes the shared file (share is still present on server 1
* PROPFIND on the received file

stable8.2 and this branch both result in:

```xml
<?xml version="1.0" encoding="utf-8"?>
<d:error xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns">
  <s:exception>Sabre\DAV\Exception\ServiceUnavailable</s:exception>
  <s:message>Storage not available</s:message>
</d:error>
```

**Comment by MorrisJobke** [comment]
> stable8.2 and this branch both result in:

Just to point out: This branch works, but it has worked before too :laughing: 

Also the propfind on the parent of the received remote share looks okay:

```xml
<?xml version="1.0" encoding="utf-8"?>
<d:multistatus xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns" xmlns:oc="http://owncloud.org/ns">
  <d:response>
    <d:href>/master/remote.php/webdav/</d:href>
    <d:propstat>
      <d:prop>
        <d:getlastmodified>Fri, 11 Dec 2015 08:38:46 GMT</d:getlastmodified>
        <d:resourcetype>
          <d:collection/>
        </d:resourcetype>
        <d:quota-used-bytes>163</d:quota-used-bytes>
        <d:quota-available-bytes>58813280256</d:quota-available-bytes>
        <d:getetag>"566a8b9629629"</d:getetag>
      </d:prop>
      <d:status>HTTP/1.1 200 OK</d:status>
    </d:propstat>
  </d:response>
  <d:response>
    <d:href>/master/remote.php/webdav/welcome.txt</d:href>
    <d:propstat>
      <d:prop>
        <d:getlastmodified>Fri, 11 Dec 2015 08:38:46 GMT</d:getlastmodified>
        <d:getcontentlength>163</d:getcontentlength>
        <d:resourcetype/>
        <d:getetag>"0c625c9035ca41e239e0da62bf9b42c2"</d:getetag>
        <d:getcontenttype>text/plain</d:getcontenttype>
      </d:prop>
      <d:status>HTTP/1.1 200 OK</d:status>
    </d:propstat>
  </d:response>
  <d:response>
    <d:href>/master/remote.php/webdav/sdf.txt</d:href>
    <d:propstat>
      <d:prop>
        <d:getlastmodified>Fri, 11 Dec 2015 08:39:00 GMT</d:getlastmodified>
        <d:getcontentlength>0</d:getcontentlength>
        <d:resourcetype/>
        <d:getetag>"e4649fd64022ae139cc3e1e29805b665"</d:getetag>
        <d:getcontenttype>text/plain</d:getcontenttype>
      </d:prop>
      <d:status>HTTP/1.1 200 OK</d:status>
    </d:propstat>
  </d:response>
</d:multistatus>
```
 
(Looks the same on stable8.2 and this branch)

**Comment by LukasReschke** [comment]
@MorrisJobke Did you set the storage to unavailable? - See https://github.com/owncloud/core/pull/21046#issuecomment-163876253

… alternatively wait a long time or so :speak_no_evil:

**Comment by MorrisJobke** [comment]
> @MorrisJobke Did you set the storage to unavailable? - See #21046 (comment)

Ah :) Let me check that

**Comment by MorrisJobke** [comment]
> Ah :) Let me check that

Yes. Still same results on stable8.2 and this branch.

**Comment by PVince81** [comment]
@MorrisJobke you need to PROPFIND with the propfind.txt file from the original ticket.
Without this, it will not trigger the bogus code path.
You need to request the "downloadURL" property.

**Comment by PVince81** [comment]
@MorrisJobke your result already shows that you didn't, it doesn't include the properties :smile: 
See the curl command from https://github.com/owncloud/core/issues/21033#issuecomment-163323692 and the file contents.

**Comment by MorrisJobke** [comment]
> @MorrisJobke your result already shows that you didn't, it doesn't include the properties  
> See the curl command from #21033 (comment) and the file contents.

@LukasReschke also explained this to me just now. I could reproduce and this fixes it.

:+1:

**Comment by LukasReschke** [comment]
:+1:

**Pull request merged**